packages/react/src/extend-theme.ts

```diff
@@ -21,45 +25,118 @@ type DeepThemeExtension<BaseTheme, ThemeType> = {
     : CloneKey<ThemeType, Key>
 }
 
-export type ThemeOverride = Partial<ChakraTheme> &
-  DeepThemeExtension<Theme, ChakraTheme>
+export type ThemeOverride<BaseTheme = Theme> = Partial<ChakraTheme> &
+  DeepThemeExtension<BaseTheme, ChakraTheme> &
+  Dict
+
+export type ThemeExtension<Override extends ThemeOverride = ThemeOverride> = (
+  themeOverride: Override,
+) => Override
+
+export type BaseThemeWithExtensions<
+  BaseTheme extends ChakraTheme,
+  Extensions extends readonly [...any]
+> = Extensions extends [infer L, ...infer R]
+  ? L extends (...args: any[]) => any
+    ? ReturnType<L> & BaseThemeWithExtensions<BaseTheme, R>
+    : L & BaseThemeWithExtensions<BaseTheme, R>
+  : BaseTheme & Extensions
 
 /**
- * Function to override or customize the Chakra UI theme conveniently
- * @param overrides - Your custom theme object overrides
- * @param baseTheme - theme to customize
+ * Function to override or customize the Chakra UI theme conveniently.
+ * First extension overrides the baseTheme and following extensions override the preceding extensions.
+ *
+ * @example with only overrides:
+ * import { extendTheme } from '@chakra-ui/react'
+ *
+ * const customTheme = extendTheme({
+ *   colors: {
+ *     brand: {
+ *       500: "#b4d455",
+ *     },
+ *   },
+ * })
+ *
+ *
+ * @example with extension:
+ * import { theme, extendTheme, withDefaultColorScheme } from '@chakra-ui/react'
+ *
+ * const customTheme = extendTheme(
+ *   {
+ *     colors: {
+ *       brand: {
+ *         500: "#b4d455",
+ *       },
+ *     },
+ *   },
+ *   withDefaultColorScheme({ colorScheme: "red" }),
+ *   theme,
+ * )
  */
 export function extendTheme<
   BaseTheme extends ChakraTheme = Theme,
-  Overrides extends ThemeOverride = ThemeOverride
+  Extensions extends (
+    | BaseTheme
+    | ThemeOverride<BaseTheme>
+    | ThemeExtension<ThemeOverride<BaseTheme>>
+  )[] = (ThemeOverride<BaseTheme> | ThemeExtension<ThemeOverride<BaseTheme>>)[]
 >(
-  overrides: Overrides,
-  baseTheme: BaseTheme = (defaultTheme as unknown) as BaseTheme,
-): BaseTheme & Overrides {
-  function customizer(
-    source: unknown,
-    override: unknown,
-    key: string,
-    object: any,
+  ...extensions: [...Extensions]
+): BaseThemeWithExtensions<BaseTheme, Extensions> {
+  const overrides = [...extensions]
+    .slice(0, extensions.length - (extensions.length > 1 ? 1 : 0))
+    .reverse()
```

Comment:
`reverse()` + `compose()` is hurting my brain trying to figure out what order the functions are called in. `extendTheme(a(), b(), c())`, what order are they called in?

Reply:
The order of overrides is from right to left, except the base:

```js
extendTheme(
  extionsionONEoverridesBase,      // <- called first
  extensionTWOoverridesONE,        // <- called second
  extensionNEXToverridesPRECEDING, // <- called third
  base,
)
```

We need to keep this function backwards compatible, so the last param needs to stay the baseTheme. 

It thought with `reverse()`  the using side is more intuitive, but I do not have a strong opinion on this.
Without reverse  the order would be:

```js
extendTheme(
  extensionNEXToverridesPRECEDING,  // <- called third
  extensionTWOoverridesONE,         // <- called second
  extionsionONEoverridesBase,       // <- called first
  base,
)
```

Reply:
I think I'm even more confused now lol, I think because I'm confused about what you mean by "right to left" and maybe FP terminology around `compose`. In FP, `compose` is RTL, meaning that it starts at the last function, executes it with the argument, and then takes that result and passes it to the function before it, and so on.

`compose(a, b, c)('hi') === a(b(c('hi')))`

The reverse (😉) of that is `pipe` aka `compose` + `reverse` which is LTR:

`pipe(a, b, c)('hi') === c(b(a('hi')))`

I don't know that I really have a preference one way or the other! It'll just be confusing to other FP-minded people if we're saying compose/RTL when we actually mean pipe/LTR 🙃

Reply:
1. I removed the `reverse` call and replaced to compose function with `pipe` and moved this function to the `utils` package.
2. The override order is "Left To Right". I think I got confused there myself when I wrote the docs.